rfcs/shared/build-system/package-public-api.md

```diff
@@ -0,0 +1,237 @@
+
+- tooling will trim the declaration from a public API surface.
+- can be exposed only from `/unstable` API surface
+- tolling will NOT TRIM the declaration from `/unstable` API surface.
+
+**Example:**
+
+- unstable api
+
+```ts
+// @filename package-a/src/unstable/index.ts
+
+/**
+ *
+ * @alpha
+ */
+export function doSomething(): void {}
+
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/unstable.d.ts
+export function doSomething(): void;
+export declare function hello(): void;
+```
+
+- stable api
+
+```ts
+// @filename package-a/src/index.ts
+
+// 🚨🚨🚨 NOTE THAT THIS IS PROBABLY A MISTAKE BUT TOOLING WILL COVER YOU
+/**
+ *
+ * @alpha
+ */
+export function doSomething(): void {}
+/**
+ *
+ * @public
+ */
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/index.d.ts
+export declare function hello(): void;
+```
+
+#### `@beta`
+
+It indicates that an API item has been released as a preview or for experimental purposes. Third parties are encouraged to try it and provide feedback. However, a “beta” API should NOT be used in production, because it may be changed or removed in a future version.
+
+- tooling will trim the declaration from a public API surface.
+- can be exposed only from `/unstable` API surface
+- tolling will NOT TRIM the declaration from `/unstable` API surface.
+
+**Example:**
+
+- unstable api
+
+```ts
+// @filename package-a/src/unstable/index.ts
+
+/**
+ *
+ * @beta
+ */
+export function doSomething(): void {}
+
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/unstable.d.ts
+export function doSomething(): void;
+export declare function hello(): void;
+```
+
+- stable api
+
+```ts
+// @filename package-a/src/index.ts
+
+// 🚨🚨🚨 NOTE THAT THIS IS PROBABLY A MISTAKE BUT TOOLING WILL COVER YOU
+/**
+ *
+ * @beta
+ */
+export function doSomething(): void {}
+/**
+ *
+ * @public
+ */
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/index.d.ts
+export declare function hello(): void;
+```
+
+#### `@public`
+
+It indicates that an API item has been officially released, and is now part of the supported contract for a package. Semver versioning is used and enforced thus the API signature cannot be changed without a MAJOR version increment.
+
+- tooling will NOT trim the declaration from a public API surface.
+- CANNOT be exposed from `/unstable` API surface
+
+**Example:**
+
+```ts
+// @filename package-a/src/index.ts
+
+/**
+ *
+ * @internal
+ */
+export function doSomething(): void {}
+/**
+ *
+ * @public
+ */
+export function hello(): void {}
+```
+
+↓↓ transforms to ↓↓
+
+```ts
+// @filename package-a/dist/index.d.ts
+export declare function hello(): void;
+```
+
+### Open questions:
+
+- should we keep `_unstable` suffix for identifiers ?
```

Comment:
I think we should keep it. Currently our composition API (e.g., `useComponent_unstable`) uses this in otherwise stable components. I could also see us adding other unstable APIs to stable components (e.g., `someProp_unstable`).

Reply:
using `_unstable` api internally is not an issue. issue is if those apis are exposed as public package surface, which is already happening in our react-components "sub" packages

![image](https://user-images.githubusercontent.com/1223799/174261280-002989ad-def9-443f-8869-0564e967658c.png)

this RFC proposes to expose this packages only from `/unstable` api of particular package, which would be enforced by tooling. This would add us following benefits:
- no churn in package apis 
- stylistic thing: no  `_unstable` churn in identifiers anymore, we can use `_` suffix maybe to simplify (although prefix is a common pattern react hooks are playing against that unfortunately )

Reply:
Let me make sure my understanding is correct:

1. `import { MenuProvider } from '@fluentui/react-menu'`
2. `import { useMenuContext } from '@fluentui/react-menu/unstable'` (currently this is exported as `useMenuContext_unstable` from the stable import path)
3. This would be mirrored in `@fluentui/react-components`

I still think it's nice to have `_unstable` in the name for identifiers like `useMenuContext_unstable`. It's just that much more explicit and it's easy enough to find and replace or codemod when `_unstable` is dropped. The extra explicitness feels nice to me because our consumers should avoid our unstable APIs if at all possible. This issue isn't a blocker for me though.

Reply:
> Let me make sure my understanding is correct:

yes thats correct 🙌

> I still think it's nice to have _unstable in the name for identifiers like useMenuContext_unstable

yeah sounds good, at least internally it will be more explicit that we are using unstable on our own. but really having a `_` would to the "trick" as well -> `useMenuContext_`